google-ads-examples/src/main/java/com/google/ads/googleads/examples/misc/SetCustomClientTimeouts.java

```diff
@@ -0,0 +1,161 @@
+      System.exit(1);
+    } catch (IOException ioe) {
+      System.err.printf("Failed to create GoogleAdsClient. Exception: %s%n", ioe);
+      System.exit(1);
+    }
+
+    try {
+      new SetCustomClientTimeouts().runExample(googleAdsClient, params.customerId);
+    } catch (GoogleAdsException gae) {
+      // GoogleAdsException is the base class for most exceptions thrown by an API request.
+      // Instances of this exception have a message and a GoogleAdsFailure that contains a
+      // collection of GoogleAdsErrors that indicate the underlying causes of the
+      // GoogleAdsException.
+      System.err.printf(
+          "Request ID %s failed due to GoogleAdsException. Underlying errors:%n",
+          gae.getRequestId());
+      int i = 0;
+      for (GoogleAdsError googleAdsError : gae.getGoogleAdsFailure().getErrorsList()) {
+        System.err.printf("  Error %d: %s%n", i++, googleAdsError);
+      }
+      System.exit(1);
+    }
+  }
+
+  /** Runs the example. */
+  private void runExample(GoogleAdsClient googleAdsClient, Long customerId) {
+    makeServerStreamingCall(googleAdsClient, customerId);
+    makeUnaryCall(googleAdsClient, customerId);
+  }
+
+  /** Makes a server streaming call using a custom client timeout. */
+  // [START SetCustomClientTimeouts]
+  private void makeServerStreamingCall(GoogleAdsClient googleAdsClient, Long customerId) {
+    StringBuilder output = new StringBuilder();
+    try (GoogleAdsServiceClient serviceClient =
+        googleAdsClient.getLatestVersion().createGoogleAdsServiceClient()) {
+      // Issues a search stream request by setting a custom client timeout
+      ServerStream<SearchGoogleAdsStreamResponse> stream =
+          serviceClient
+              .searchStreamCallable()
+              .call(
+                  SearchGoogleAdsStreamRequest.newBuilder()
+                      .setCustomerId(String.valueOf(customerId))
+                      .setQuery("SELECT campaign.id FROM campaign")
+                      .build(),
+                  // Sets the timeout to use.
+                  GrpcCallContext.createDefault()
+                      .withTimeout(Duration.ofMillis(CLIENT_TIMEOUT_MILLIS)));
+      // Iterates over all rows in all messages and collects the campaign IDs.
+      for (SearchGoogleAdsStreamResponse response : stream) {
+        for (GoogleAdsRow googleAdsRow : response.getResultsList()) {
+          output.append(" ").append(googleAdsRow.getCampaign().getId());
+        }
+      }
+      System.out.println("The server streaming call completed before the timeout");
+    } catch (GoogleAdsException ex) {
+      if (ex.getStatusCode().getCode() == Code.DEADLINE_EXCEEDED) {
+        System.out.println("The server streaming call did not complete before the timeout.");
+      } else {
+        // Bubbles up if the exception is not about timeout.
+        throw ex;
+      }
+    }
+    System.out.printf("All campaign IDs retrieved: %s.%n", output.toString());
```

Comment:
This should be in a finally block so that it prints even if an error (typically a timeout here) is reached.

Reply:
I assumed this was only supposed to print for the successful case. Are you saying that this message is meant to tell the user "I managed to retrieve these campaign IDs...", even if an exception occurred before reaching the end of the stream?

Reply:
Correct. The idea is that the stream can receive results before a timeout (or any other exception) arises.

Reply:
Done